addons/base/views.py

```diff
@@ -0,0 +1,742 @@
+        # Central Authentication Server OAuth Bearer Token
+        authorization = request.headers.get('Authorization')
+        if authorization and authorization.startswith('Bearer '):
+            client = cas.get_client()
+            try:
+                access_token = cas.parse_auth_header(authorization)
+                cas_resp = client.profile(access_token)
+            except cas.CasError as err:
+                sentry.log_exception()
+                # NOTE: We assume that the request is an AJAX request
+                return json_renderer(err)
+            if cas_resp.authenticated:
+                auth.user = User.load(cas_resp.user)
+
+    try:
+        data = jwt.decode(
+            jwe.decrypt(request.args.get('payload', '').encode('utf-8'), WATERBUTLER_JWE_KEY),
+            settings.WATERBUTLER_JWT_SECRET,
+            options={'require_exp': True},
+            algorithm=settings.WATERBUTLER_JWT_ALGORITHM
+        )['data']
+    except (jwt.InvalidTokenError, KeyError):
+        raise HTTPError(httplib.FORBIDDEN)
+
+    if not auth.user:
+        auth.user = User.from_cookie(data.get('cookie', ''))
+
+    try:
+        action = data['action']
+        node_id = data['nid']
+        provider_name = data['provider']
+    except KeyError:
+        raise HTTPError(httplib.BAD_REQUEST)
+
+    node = Node.load(node_id)
+    if not node:
+        raise HTTPError(httplib.NOT_FOUND)
+
+    check_access(node, auth, action, cas_resp)
+
+    provider_settings = node.get_addon(provider_name)
+    if not provider_settings:
+        raise HTTPError(httplib.BAD_REQUEST)
+
+    try:
+        credentials = provider_settings.serialize_waterbutler_credentials()
+        waterbutler_settings = provider_settings.serialize_waterbutler_settings()
+    except exceptions.AddonError:
+        log_exception()
+        raise HTTPError(httplib.BAD_REQUEST)
+
+    return {'payload': jwe.encrypt(jwt.encode({
+        'exp': timezone.now() + datetime.timedelta(seconds=settings.WATERBUTLER_JWT_EXPIRATION),
+        'data': {
+            'auth': make_auth(auth.user),  # A waterbutler auth dict not an Auth object
+            'credentials': credentials,
+            'settings': waterbutler_settings,
+            'callback_url': node.api_url_for(
+                ('create_waterbutler_log' if not node.is_registration else 'registration_callbacks'),
+                _absolute=True,
+            ),
+        }
+    }, settings.WATERBUTLER_JWT_SECRET, algorithm=settings.WATERBUTLER_JWT_ALGORITHM), WATERBUTLER_JWE_KEY)}
+
+
+LOG_ACTION_MAP = {
+    'move': NodeLog.FILE_MOVED,
+    'copy': NodeLog.FILE_COPIED,
+    'rename': NodeLog.FILE_RENAMED,
+    'create': NodeLog.FILE_ADDED,
+    'update': NodeLog.FILE_UPDATED,
+    'delete': NodeLog.FILE_REMOVED,
+    'create_folder': NodeLog.FOLDER_CREATED,
+}
+
+
+@must_be_signed
+@no_auto_transaction
+@must_be_valid_project
+def create_waterbutler_log(payload, **kwargs):
+    with TokuTransaction():
```

Comment:
No, it isn't; I've been switching these to `atomic.transaction()`, but I missed this one. I'll change this over, but it wouldn't be a bad idea to patch it. I'll do this in a separate PR.